### ISSUE TITLE
Using a different serializer

### ISSUE THREAD

**Issue by ahychu84** [Abc-Arbitrage/Zebus#83]
This is more of a question than an issue but I was wondering whether it would be possible to swap out Protobuf with a different serializer/deserializer?

**Comment by ltrzesniewski** [comment]
That should be doable but you'd need to make an incompatible fork of the library. We have no plans to change the serialization format as Protobuf works nicely for our purposes.

Protobuf is integrated quite deeply and some optimizations are implemented around it, for instance Zebus uses knowledge of the format to patch some bytes in a message in order to avoid having to serialize a message multiple times.

**Comment by ocoanet** [comment]
Well, Protobuf is used for two different parts:

- For user message serialization.

This is supposed to be swappable by another serialization format. You can implement and provide your own `IMessageSerializer`.

```cs
namespace Abc.Zebus.Serialization
{
    public interface IMessageSerializer
    {
        Stream Serialize(IMessage message);
        IMessage Deserialize(MessageTypeId messageTypeId, Stream stream);
    }
}
```

To be honest I found a couple of cases where Protobuf is used directly instead of `IMessageSerializer` on the codebase but I can fix that really quickly.

- For transport message serialization.

This is supposed to be an implementation detail. The format and protocol could be changed in future versions. You are not supposed to rely on this and you cannot change it.

I hope this will help.

**Comment by ocoanet** [comment]
@ltrzesniewski https://github.com/Abc-Arbitrage/Zebus/commit/a9e8a6803372217b8c34ef6f7991e43608a2be57 ouch 😄

**Comment by ltrzesniewski** [comment]
Oops! 😰
TBH I never thought of `IMessageSerializer` as something that was overridable. I'll fix that.

**Comment by ahychu84** [comment]
Awesome thanks for all the help everyone

**Comment by ahychu84** [comment]
@ltrzesniewski When abouts do you think a fix will be pushed? Just curious when I can start playing around with this code

**Comment by ltrzesniewski** [comment]
@ahychu84 I'm not sure, I'll try to find some time to fix it.

But you can already start playing around if you want to. The problematic code that @ocoanet found is only used in a very narrow edge case, and should basically be a no-op if you don't use Protobuf.

**Comment by ahychu84** [comment]
@ltrzesniewski Thanks for the quick response. Looking into the code it looks like protobufs are baked in to the actual protocol which is being used to transport messages between its peers. I will play around protobufs and see how it works. I have used it previously with some other implementations and found it very cumbersome so perhaps this experience will be different.

**Comment by ltrzesniewski** [comment]
> Looking into the code it looks like protobufs are baked in to the actual protocol which is being used to transport messages between its peers.

Yes, that's what @ocoanet mentioned in the "For transport message serialization." part of his comment. It's an implementation detail, and the code has some optimizations which rely on knowledge of the format.

> I have used it previously with some other implementations and found it very cumbersome so perhaps this experience will be different.

Protobuf works very well for us. I don't know why you consider it cumbersome but if it's because of the annotations you need to add to the message types, take a look at the [Zebus.MessageDsl](https://github.com/Abc-Arbitrage/Zebus.MessageDsl) package which makes this problem go away.

**Comment by ltrzesniewski** [comment]
Reopening this to remind me of the cloning issue.

**Comment by ltrzesniewski** [comment]
I just released v3.3.0 with the fix, see #84 for more info.
